cpp/memilio/io/io.h

```diff
@@ -17,131 +17,22 @@
- * expected to store errors. During serialization, `add_...` operations fail without returning errors, but the error is stored 
- * in the IOObject and subsequent calls are usually no-ops. During deserialization, the values produced must usually be used or 
- * inspected, so `expect_...` operations return an IOResult. The `apply` utility function provides a simple way to inspect the result 
- * of multiple `expect_...` operations and use the values if all are succesful. See the documentation of `IOStatus`, `IOResult`
- * and `apply` below for more details.
- * 
- * Implementing serialization for a new type:
- * ------------------------------------------
- * Serialization of a new type T can be customized by providing _either_ member functions `serialize` and `deserialize` _or_ free functions 
- * `serialize_internal` and `deserialize_internal`.
- * 
- * The `void serialize(IOContext& io)` member function takes an IO context and uses `create_object` and `add_...` operations
- *  to add data. The static `IOResult<T> deserialize(IOContext& io)` member function takes an IO context and uses `expect_...` 
- * operations to retrieve the data. The `apply` utility function can be used to inspect the result of the `expect_...`
- * operations and construct the object of type T.
- * E.g.:
- * ```
- * struct Foo {
- *   int i;
- *   template<class IOContext>
- *   void serialize(IOContext& io) {
- *     auto obj = io.create_object("Foo");
- *     obj.add_element("i", i);
- *   }
- *   template<class IOContext>
- *   static IOResult<Foo> deserialize(IOContext& io) {
- *     auto obj = io.expect_object("Foo");
- *     auto i_result = obj.expect_element("i", epi::Tag<int>{});
- *     return epi::apply(io, [](auto&& i) { return Foo{i}; }, i_result);
- *   }
- * };
- * ```
- * 
- * The free functions `serialize_internal` and `deserialize_internal` must be found with argument dependent lookup (ADL). 
- * They can be used if no member function should or can be added to the type. See the code below for examples where 
- * this was done for, e.g., Eigen3 matrices and STL containers.
- * 
- * Adding a new serialization format:
- * ----------------------------------
- * Implement concepts IOContext and IOObject that provide the operations listed above. Your implemenation should handle
- * all built in types as well as std::string. It may handle other types (e.g., STL containers) as well if it can do so
- * more efficiently than the provided general free functions. See the JsonSerializer implementation in the epidemiology_io
- * library for an example.
- */
 
 #ifndef EPI_UTILS_IO_H
 #define EPI_UTILS_IO_H
 
-#include "epidemiology/utils/metaprogramming.h"
-#include "epidemiology/utils/eigen_util.h"
+#include "memilio/utils/metaprogramming.h"
+#include "memilio/utils/compiler_diagnostics.h"
+#include "memilio/math/eigen_util.h"
+GCC_CLANG_DIAGNOSTIC(push)
+GCC_CLANG_DIAGNOSTIC(ignored "-Wc++17-extensions")
```

Comment:
boost checks for specific features being available instead of checking the version of the standard. In this case, the `nodiscard` attribute was supported by gcc and clang as an extension long before c++17 came out, and it's really helpful to detect some errors. I think boost can be trusted to properly guard these extensions if they are not available, so I just disabled the warning here (but not globally). I don't think anything bad can happen if we forget this, worst case it becomes obsolete if we upgrade to c++17.